jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/MainClassResolver.java

```diff
@@ -40,83 +39,59 @@
-      // don't error if not found in class files.
-      try {
-        mainClass = findMainClassInClassFiles(projectProperties);
-
-      } catch (MainClassInferenceException ignored) {
-        // Fallback to using the mainClass found in projectProperties.
-      }
-    }
-
-    Preconditions.checkNotNull(mainClass);
-    if (!isValidJavaClass(mainClass)) {
-      projectProperties.log(LogEvent.warn("'mainClass' is not a valid Java class : " + mainClass));
-    }
-
-    return mainClass;
-  }
-
-  /**
-   * Checks if a string is a valid Java class name.
-   *
-   * @param className the class name to check
-   * @return {@code true} if {@code className} is a valid Java class name; {@code false} otherwise
-   */
-  @VisibleForTesting
-  static boolean isValidJavaClass(String className) {
-    for (String part : Splitter.on('.').split(className)) {
-      if (!SourceVersion.isIdentifier(part)) {
-        return false;
+    if (configuredMainClass != null) {
+      if (isValidJavaClass(configuredMainClass)) {
+        return configuredMainClass;
       }
+      throw new MainClassInferenceException(
+          HelpfulSuggestions.forMainClassNotFound(
+              "'mainClass' configured in "
+                  + projectProperties.getPluginName()
+                  + " is not a valid Java class: "
+                  + configuredMainClass,
+              projectProperties.getPluginName()));
     }
-    return true;
-  }
 
-  @Nullable
-  private static String getMainClassFromJar(ProjectProperties projectProperties) {
     projectProperties.log(
         LogEvent.info(
             "Searching for main class... Add a 'mainClass' configuration to '"
                 + projectProperties.getPluginName()
                 + "' to improve build speed."));
-    return projectProperties.getMainClassFromJar();
-  }
 
-  private static String findMainClassInClassFiles(ProjectProperties projectProperties)
-      throws MainClassInferenceException, IOException {
+    String mainClassFromJar = projectProperties.getMainClassFromJar();
+    if (mainClassFromJar != null && isValidJavaClass(mainClassFromJar)) {
+      return mainClassFromJar;
+    }
+
+    if (mainClassFromJar != null) {
+      projectProperties.log(
+          LogEvent.warn(
+              "'mainClass' configured in "
+                  + projectProperties.getJarPluginName()
+                  + " is not a valid Java class: "
+                  + mainClassFromJar));
+    }
     projectProperties.log(
-        LogEvent.debug(
-            "Could not find a valid main class specified in "
+        LogEvent.info(
+            "Could not find a valid main class from "
                 + projectProperties.getJarPluginName()
-                + "; attempting to infer main class."));
+                + "; looking into all class files to infer main class."));
 
-    MainClassFinder.Result mainClassFinderResult =
+    MainClassFinder.Result result =
```

Comment:
I actually found mainClassFinderResult to be easier to read/follow. But no strong feelings here.